crates/noirc_evaluator/src/ssa/acir_gen/constraints.rs

```diff
@@ -0,0 +1,678 @@
+use crate::{
+    errors::RuntimeErrorKind,
+    ssa::acir_gen::{const_from_expression, expression_from_witness, expression_to_witness},
+    Evaluator,
+};
+use acvm::{
+    acir::{
+        circuit::{
+            directives::Directive,
+            opcodes::{BlackBoxFuncCall, FunctionInput, Opcode as AcirOpcode},
+        },
+        native_types::{Expression, Linear, Witness},
+    },
+    FieldElement,
+};
+use num_bigint::BigUint;
+use num_traits::One;
+use std::{
+    cmp::Ordering,
+    ops::{Mul, Neg},
+};
+
+// Code in this file, will generate constraints without
+// using only the Evaluator and ACIR Expression types
+
+// This interface abstracts away the functionality needed
+// by the `Evaluator` to create constraints.
+pub trait ACIRState {
```

Comment:
Why do we need this trait if it will only be implemented for the Evaluator? If we want method syntax we can `impl Evaluator` directly.

Reply:
The plan is to move this trait and many of the functions in constraints into ACVM stdlib and then we impl this for Evaluator here.